ambry-server/src/integration-test/java/com.github.ambry.server/RouterServerTestFramework.java

```diff
@@ -0,0 +1,476 @@
+import com.github.ambry.router.CoordinatorBackedRouterMetrics;
+import com.github.ambry.router.NonBlockingRouterFactory;
+import com.github.ambry.router.ReadableStreamChannel;
+import com.github.ambry.router.Router;
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Properties;
+import java.util.Queue;
+import java.util.Random;
+import java.util.concurrent.CountDownLatch;
+import java.util.concurrent.Future;
+import org.junit.Assert;
+
+
+class RouterServerTestFramework {
+  static final int CHUNK_SIZE = 1024 * 1024;
+  private final MockClusterMap clusterMap;
+  private final MockNotificationSystem notificationSystem;
+  private final Router nonBlockingRouter;
+  private final Router coordinatorBackedRouter;
+
+  RouterServerTestFramework(Properties routerProps, MockCluster cluster, MockNotificationSystem notificationSystem)
+      throws Exception {
+    this.clusterMap = cluster.getClusterMap();
+    this.notificationSystem = notificationSystem;
+
+    VerifiableProperties routerVerifiableProps = new VerifiableProperties(routerProps);
+    this.nonBlockingRouter =
+        new NonBlockingRouterFactory(routerVerifiableProps, clusterMap, notificationSystem).getRouter();
+
+    RouterConfig routerConfig = new RouterConfig(routerVerifiableProps);
+    CoordinatorBackedRouterMetrics coordinatorBackedRouterMetrics =
+        new CoordinatorBackedRouterMetrics(clusterMap.getMetricRegistry());
+    AmbryCoordinator coordinator = new HelperCoordinator(routerVerifiableProps, clusterMap);
+    this.coordinatorBackedRouter =
+        new CoordinatorBackedRouter(routerConfig, coordinatorBackedRouterMetrics, coordinator);
+  }
+
+  void cleanup()
+      throws IOException {
+    if (nonBlockingRouter != null) {
+      nonBlockingRouter.close();
+    }
+    if (coordinatorBackedRouter != null) {
+      coordinatorBackedRouter.close();
+    }
+  }
+
+  void checkOperationChains(List<OperationInfo> opInfos)
+      throws Exception {
+    Map<PartitionId, Integer> partitionCount = new HashMap<>();
+    double blobsPut = 0;
+    for (OperationInfo opInfo : opInfos) {
+      opInfo.latch.await();
+      synchronized (opInfo.futures) {
+        for (TestFuture future : opInfo.futures) {
+          future.check();
+        }
+      }
+      if (opInfo.blobId != null) {
+        blobsPut++;
+        PartitionId partitionId = new BlobId(opInfo.blobId, clusterMap).getPartition();
+        int count = partitionCount.containsKey(partitionId) ? partitionCount.get(partitionId) : 0;
+        partitionCount.put(partitionId, count + 1);
+      }
+    }
+    double blobBalanceThreshold = 3.0 * Math.ceil(blobsPut / (double) clusterMap.getWritablePartitionIds().size());
+    for (Map.Entry<PartitionId, Integer> entry : partitionCount.entrySet()) {
+      Assert.assertTrue("Number of blobs is " + entry.getValue() + " on partition: " + entry.getKey()
```

Comment:
This threshold is pretty conservative (3 times expected value).  Also, the partition chosen should be assigned randomly so ideally it should be close to balanced.